Thank you for the review!

The WPE build failure:

```
[2146/2194] Generating WPEWebProcessExtension-2.0.gir
FAILED: WPEWebProcessExtension-2.0.gir /app/webkit/WebKitBuild/WPE/Release/WPEWebProcessExtension-2.0.gir 
cd /app/webkit && /usr/bin/cmake -E env CC=/usr/bin/cc "CFLAGS=-fdiagnostics-color=always -Wextra -Wall -pipe -fmax-errors=20 -Wno-expansion-to-defined -Wno-psabi -Wno-misleading-indentation -Wno-maybe-uninitialized -Wundef -Wpointer-arith -Wmissing-format-attribute -Wformat-security -Wcast-align -Wno-tautological-compare -fdebug-types-section -fno-strict-aliasing -fno-exceptions -ffunction-sections -fdata-sections" /usr/bin/g-ir-scanner --quiet --warn-all --warn-error --no-libtool --output=/app/webkit/WebKitBuild/WPE/Release/WPEWebProcessExtension-2.0.gir --library=WPEWebKit-2.0 --library-path=/app/webkit/WebKitBuild/WPE/Release/lib --namespace=WPEWebProcessExtension --nsversion=2.0 --c-include=wpe/wpe-web-process-extension.h --identifier-prefix=WebKit --symbol-prefix=webkit --pkg-export=wpe-web-process-extension-2.0 -DBUILDING_WebKit -DBUILDING_WITH_CMAKE=1 -DBUILDING_WEBKIT=1 -DHAVE_CONFIG_H=1 -DPAS_BMALLOC=1 -DBUILDING_WPE__=1 -DGETTEXT_PACKAGE=\"WPE\" -DJSC_GLIB_API_ENABLED -DBWRAP_EXECUTABLE=\"/usr/bin/bwrap\" -DDBUS_PROXY_EXECUTABLE=\"/usr/bin/xdg-dbus-proxy\" -DLIBDIR=\"/usr/local/lib\" -DPKGLIBDIR=\"/usr/local/lib/wpe-webkit-2.0\" -DPKGLIBEXECDIR=\"/usr/local/libexec/wpe-webkit-2.0\" -DDATADIR=\"/usr/local/share\" -DLOCALEDIR=\"/usr/local/share/locale\" -DSK_DISABLE_LEGACY_GL_MAKE_NATIVE_INTERFACE -DSK_DISABLE_LEGACY_IMAGE_READBUFFER -DSK_DISABLE_LEGACY_INIT_DECODERS -DSK_DISABLE_LEGACY_PNG_WRITEBUFFER -DSK_DISABLE_TRACING -DSK_DISABLE_WEBGL_INTERFACE -DSK_ENABLE_PRECOMPILE -DSK_R32_SHIFT=16 -DSK_RELEASE -DSK_ASSUME_GL=0 -DSK_ASSUME_GL_ES=1 -DSK_ASSUME_WEBGL=0 -DSTATICALLY_LINKED_WITH_bmalloc -DSTATICALLY_LINKED_WITH_JavaScriptCore -DSTATICALLY_LINKED_WITH_WTF -DSTATICALLY_LINKED_WITH_PAL -DSTATICALLY_LINKED_WITH_WebCore -DUSE_SYSTEM_EGL -I/app/webkit/Source/ThirdParty/skia -I/app/webkit/WebKitBuild/WPE/Release/Skia/Headers -I/usr/include/wpe-1.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include --sources-top-dirs=/app/webkit --include-uninstalled=/app/webkit/WebKitBuild/WPE/Release/WPEJavaScriptCore-2.0.gir --library=WPEWebKit-2.0 --include=Soup-3.0 --pkg=libsoup-3.0 -I/app/webkit/WebKitBuild/WPE/Release/DerivedSources/ForwardingHeaders -I/app/webkit/WebKitBuild/WPE/Release/DerivedSources/ForwardingHeaders -I/app/webkit/WebKitBuild/WPE/Release/DerivedSources/ForwardingHeaders/wpe -I/app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit -I/app/webkit/WebKitBuild/WPE/Release/JavaScriptCoreGLib/Headers -I/app/webkit/WebKitBuild/WPE/Release/JavaScriptCoreGLib/DerivedSources /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitWebProcessEnumTypes.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitConsoleMessage.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitFrame.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitScriptWorld.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitWebEditor.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitWebFormManager.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitWebHitTestResult.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitWebPage.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/webkit-web-process-extension.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitWebProcessExtension.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitContextMenu.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitContextMenuActions.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitContextMenuItem.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitHitTestResult.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitUserMessage.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitURIRequest.h /app/webkit/WebKitBuild/WPE/Release/DerivedSources/WebKit/wpe/WebKitURIResponse.h /app/webkit/Source/WebKit/Shared/API/glib/WebKitContextMenu.cpp /app/webkit/Source/WebKit/Shared/API/glib/WebKitContextMenuItem.cpp /app/webkit/Source/WebKit/Shared/API/glib/WebKitHitTestResult.cpp /app/webkit/Source/WebKit/Shared/API/glib/WebKitUserMessage.cpp /app/webkit/Source/WebKit/Shared/API/glib/WebKitURIRequest.cpp /app/webkit/Source/WebKit/Shared/API/glib/WebKitURIResponse.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitConsoleMessage.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitFrame.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitInjectedBundleMain.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitScriptWorld.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebEditor.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebExtension.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebFormManager.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebHitTestResult.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebProcessExtension.cpp /app/webkit/Source/WebKit/WebProcess/InjectedBundle/API/glib/WebProcessExtensionManager.cpp
Traceback (most recent call last):
  File "/usr/lib/python3.11/shutil.py", line 825, in move
    os.rename(src, real_dst)
OSError: [Errno 18] Invalid cross-device link: '/tmp/g-ir-scanner-cache-z84sozfz' -> '/home/ews/.cache/g-ir-scanner/8d2867c379cbf3794f3e1417464449db0efabfee'
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/bin/g-ir-scanner", line 99, in <module>
    sys.exit(scanner_main(sys.argv))
             ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/scannermain.py", line 590, in scanner_main
    transformer = create_transformer(namespace, options)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/scannermain.py", line 411, in create_transformer
    transformer.register_include_uninstalled(include_path)
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/transformer.py", line 152, in register_include_uninstalled
    self._parse_include(include_path, uninstalled=True)
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/transformer.py", line 225, in _parse_include
    self._cachestore.store(filename, parser)
  File "/usr/lib/x86_64-linux-gnu/gobject-introspection/giscanner/cachestore.py", line 153, in store
    shutil.move(tmp_filename, store_filename)
  File "/usr/lib/python3.11/shutil.py", line 845, in move
    copy_function(src, real_dst)
  File "/usr/lib/python3.11/shutil.py", line 437, in copy2
    copystat(src, dst, follow_symlinks=follow_symlinks)
  File "/usr/lib/python3.11/shutil.py", line 382, in copystat
    lookup("chmod")(dst, mode, follow_symlinks=follow)
FileNotFoundError: [Errno 2] No such file or directory: '/home/ews/.cache/g-ir-scanner/8d2867c379cbf3794f3e1417464449db0efabfee'
```

...seems unrelated at first glance. I've kicked off a rebuild just in case.

--- a/Source/WebKit/UIProcess/ios/DragDropInteractionState.h
+++ b/Source/WebKit/UIProcess/ios/DragDropInteractionState.h
@@ -45,9 +45,6 @@ struct TextIndicatorData;
 
 namespace WebKit {
 
-using DragPreparationBlock = BOOL(^)(void);
-using AddDragItemBlock = BOOL(^)(NSArray<UIDragItem *> *);
-
 struct DragSourceState {
     OptionSet<WebCore::DragSourceAction> action;
     CGRect dragPreviewFrameInRootViewCoordinates { CGRectZero };
@@ -71,7 +68,7 @@ public:
     bool anyActiveDragSourceContainsSelection() const;
 
     // These helper methods are unique to UIDragInteraction.
-    void prepareForDragSession(id<UIDragSession>, DragPreparationBlock);
+    void prepareForDragSession(id <UIDragSession>, dispatch_block_t completionHandler);
     void dragSessionWillBegin();
     void stageDragItem(const WebCore::DragItem&, UIImage *);
     bool hasStagedDragSource() const;
@@ -82,7 +79,7 @@ public:
     UITargetedDragPreview *previewForCancelling(UIDragItem *, UIView *contentView, UIView *previewContainer);
     void dragSessionWillDelaySetDownAnimation(dispatch_block_t completion);
     bool shouldRequestAdditionalItemForDragSession(id <UIDragSession>) const;
-    void dragSessionWillRequestAdditionalItem(AddDragItemBlock);
+    void dragSessionWillRequestAdditionalItem(void (^completion)(NSArray <UIDragItem *> *));
 
     void dropSessionDidEnterOrUpdate(id <UIDropSession>, const WebCore::DragData&);
     void dropSessionDidExit() { m_dropSession = nil; }
@@ -95,8 +92,8 @@ public:
     bool isPerformingDrop() const { return m_isPerformingDrop; }
     id<UIDragSession> dragSession() const { return m_dragSession.get(); }
     id<UIDropSession> dropSession() const { return m_dropSession.get(); }
-    BlockPtr<BOOL()> takeDragStartCompletionBlock() { return std::exchange(m_dragStartCompletionBlock, { }); }
-    BlockPtr<BOOL(NSArray<UIDragItem *> *)> takeAddDragItemCompletionBlock() { return std::exchange(m_addDragItemCompletionBlock, { }); }
+    BlockPtr<void()> takeDragStartCompletionBlock() { return WTFMove(m_dragStartCompletionBlock); }
+    BlockPtr<void(NSArray<UIDragItem *> *)> takeAddDragItemCompletionBlock() { return WTFMove(m_addDragItemCompletionBlock); }
     RetainPtr<UIView> takePreviewViewForDragCancel() { return std::exchange(m_previewViewForDragCancel, { }); }
 
     void addDefaultDropPreview(UIDragItem *, UITargetedDragPreview *);
@@ -117,8 +114,8 @@ private:
     bool m_isPerformingDrop { false };
     RetainPtr<id <UIDragSession>> m_dragSession;
     RetainPtr<id <UIDropSession>> m_dropSession;
-    BlockPtr<BOOL()> m_dragStartCompletionBlock;
-    BlockPtr<BOOL(NSArray<UIDragItem *> *)> m_addDragItemCompletionBlock;
+    BlockPtr<void()> m_dragStartCompletionBlock;
+    BlockPtr<void(NSArray<UIDragItem *> *)> m_addDragItemCompletionBlock;
     RetainPtr<UIView> m_previewViewForDragCancel;
 
     std::optional<DragSourceState> m_stagedDragSource;

--- a/Source/WebKit/UIProcess/ios/DragDropInteractionState.mm
+++ b/Source/WebKit/UIProcess/ios/DragDropInteractionState.mm
@@ -186,7 +186,7 @@ bool DragDropInteractionState::anyActiveDragSourceContainsSelection() const
     return false;
 }
 
-void DragDropInteractionState::prepareForDragSession(id<UIDragSession> session, DragPreparationBlock completionHandler)
+void DragDropInteractionState::prepareForDragSession(id<UIDragSession> session, dispatch_block_t completionHandler)
 {
     m_dragSession = session;
     m_dragStartCompletionBlock = completionHandler;
@@ -332,7 +332,7 @@ bool DragDropInteractionState::shouldRequestAdditionalItemForDragSession(id <UID
     return m_dragSession == session && !m_addDragItemCompletionBlock && !m_dragStartCompletionBlock;
 }
 
-void DragDropInteractionState::dragSessionWillRequestAdditionalItem(AddDragItemBlock completion)
+void DragDropInteractionState::dragSessionWillRequestAdditionalItem(void (^completion)(NSArray <UIDragItem *> *))
 {
     clearStagedDragSource();
     m_addDragItemCompletionBlock = completion;

--- a/Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm
+++ b/Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm
@@ -10049,7 +10049,7 @@ static BOOL shouldEnableDragInteractionForPolicy(_WKDragInteractionPolicy policy
 
 - (void)_didHandleDragStartRequest:(BOOL)started
 {
-    auto savedCompletionBlock = _dragDropInteractionState.takeDragStartCompletionBlock();
+    BlockPtr<void()> savedCompletionBlock = _dragDropInteractionState.takeDragStartCompletionBlock();
     ASSERT(savedCompletionBlock);
 
     RELEASE_LOG(DragAndDrop, "Handling drag start request (started: %d, completion block: %p)", started, savedCompletionBlock.get());
@@ -10525,11 +10525,18 @@ static Vector<WebCore::IntSize> sizesOfPlaceholderElementsToInsertWhenDroppingIt
     return dataOwner;
 }
 
-#if USE(BROWSERENGINEKIT)
+- (void)_dragInteraction:(UIDragInteraction *)interaction itemsForAddingToSession:(id<UIDragSession>)session withTouchAtPoint:(CGPoint)point completion:(void(^)(NSArray<UIDragItem *> *))completion
+{
+    if (!_dragDropInteractionState.shouldRequestAdditionalItemForDragSession(session)) {
+        completion(@[ ]);
+        return;
+    }
 
-#pragma mark - BEDragInteractionDelegate
+    _dragDropInteractionState.dragSessionWillRequestAdditionalItem(completion);
+    _page->requestAdditionalItemsForDragSession(WebCore::roundedIntPoint(point), WebCore::roundedIntPoint(point), self._allowedDragSourceActions);
+}
 
-- (void)dragInteraction:(BEDragInteraction *)interaction prepareDragSession:(id<UIDragSession>)session completion:(BOOL(^)(void))completion
+- (void)_dragInteraction:(UIDragInteraction *)interaction prepareForSession:(id<UIDragSession>)session completion:(dispatch_block_t)completion
 {
     RELEASE_LOG(DragAndDrop, "Preparing for drag session: %p", session);
     if (self.currentDragOrDropSession) {
@@ -10560,19 +10567,6 @@ static Vector<WebCore::IntSize> sizesOfPlaceholderElementsToInsertWhenDroppingIt
     prepareForSession(WebKit::ProceedWithTextSelectionInImage::No);
 #endif
 }
-
-- (void)dragInteraction:(BEDragInteraction *)interaction itemsForAddingToSession:(id<UIDragSession>)session forTouchAtPoint:(CGPoint)point completion:(BOOL(^)(NSArray<UIDragItem *> *))completion
-{
-    if (!_dragDropInteractionState.shouldRequestAdditionalItemForDragSession(session)) {
-        completion(@[ ]);
-        return;
-    }
-
-    _dragDropInteractionState.dragSessionWillRequestAdditionalItem(completion);
-    _page->requestAdditionalItemsForDragSession(WebCore::roundedIntPoint(point), WebCore::roundedIntPoint(point), self._allowedDragSourceActions);
-}
-
-#endif // USE(BROWSERENGINEKIT)
 
 - (NSArray<UIDragItem *> *)dragInteraction:(UIDragInteraction *)interaction itemsForBeginningSession:(id<UIDragSession>)session
 {
@@ -10705,6 +10699,26 @@ static Vector<WebCore::IntSize> sizesOfPlaceholderElementsToInsertWhenDroppingIt
 {
     [existingLocalDragSessionContext(session) cleanUpTemporaryDirectories];
 }
+
+#if USE(BROWSERENGINEKIT)
+
+#pragma mark - BEDragInteractionDelegate
+
+- (void)dragInteraction:(BEDragInteraction *)interaction prepareDragSession:(id<UIDragSession>)session completion:(BOOL(^)(void))completion
+{
+    [self _dragInteraction:interaction prepareForSession:session completion:[completion = makeBlockPtr(completion)] {
+        completion();
+    }];
+}
+
+- (void)dragInteraction:(BEDragInteraction *)interaction itemsForAddingToSession:(id<UIDragSession>)session forTouchAtPoint:(CGPoint)point completion:(BOOL(^)(NSArray<UIDragItem *> *))completion
+{
+    [self _dragInteraction:interaction itemsForAddingToSession:session withTouchAtPoint:point completion:[completion = makeBlockPtr(completion)](NSArray<UIDragItem *> *items) {
+        completion(items);
+    }];
+}
+
+#endif // USE(BROWSERENGINEKIT)
 
 #pragma mark - UIDropInteractionDelegate
 

--- a/Tools/TestRunnerShared/spi/UIKitSPIForTesting.h
+++ b/Tools/TestRunnerShared/spi/UIKitSPIForTesting.h
@@ -141,6 +141,12 @@ WTF_EXTERN_C_END
 @interface UITextInputTraits : NSObject <UITextInputTraits, UITextInputTraits_Private, NSCopying>
 @end
 
+@protocol UIDragInteractionDelegate_ForWebKitOnly<UIDragInteractionDelegate>
+@optional
+- (void)_dragInteraction:(UIDragInteraction *)interaction prepareForSession:(id<UIDragSession>)session completion:(void(^)(void))completion;
+- (void)_dragInteraction:(UIDragInteraction *)interaction itemsForAddingToSession:(id<UIDragSession>)session withTouchAtPoint:(CGPoint)point completion:(void(^)(NSArray<UIDragItem *> *))completion;
+@end
+
 @class WebEvent;
 
 @protocol UITextInputPrivate <UITextInput, UITextInputTraits_Private>

--- a/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm
+++ b/Tools/TestWebKitAPI/Tests/ios/DragAndDropTestsIOS.mm
@@ -35,7 +35,6 @@
 #import "TestWKWebView.h"
 #import "UIKitSPIForTesting.h"
 #import "WKWebViewConfigurationExtras.h"
-#import <BrowserEngineKit/BrowserEngineKit.h>
 #import <Contacts/Contacts.h>
 #import <MapKit/MapKit.h>
 #import <MobileCoreServices/MobileCoreServices.h>
@@ -50,6 +49,10 @@
 #import <WebKit/_WKProcessPoolConfiguration.h>
 #import <wtf/Seconds.h>
 #import <wtf/SoftLinking.h>
+
+#if USE(BROWSERENGINEKIT)
+#import <BrowserEngineKit/BrowserEngineKit.h>
+#endif
 
 SOFT_LINK_FRAMEWORK(Contacts)
 SOFT_LINK_CLASS(Contacts, CNMutableContact)
@@ -1579,9 +1582,13 @@ TEST(DragAndDropTests, UnresponsivePageDoesNotHangUI)
 
     // The test passes if we can prepare for a drag session without timing out.
     auto dragSession = adoptNS([[MockDragSession alloc] init]);
+#if USE(BROWSERENGINEKIT)
     [[webView dragInteractionDelegate] dragInteraction:[webView dragInteraction] prepareDragSession:dragSession.get() completion:^{
         return NO;
     }];
+#else
+    [(id<UIDragInteractionDelegate_ForWebKitOnly>)[webView dragInteractionDelegate] _dragInteraction:[webView dragInteraction] prepareForSession:dragSession.get() completion:^{ }];
+#endif
 }
 
 TEST(DragAndDropTests, WebItemProviderPasteboardLoading)

--- a/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h
+++ b/Tools/TestWebKitAPI/cocoa/DragAndDropSimulator.h
@@ -72,9 +72,14 @@ typedef NSDictionary<NSNumber *, NSValue *> *ProgressToCGPointValueMap;
 
 @interface WKWebView (DragAndDropTesting)
 - (id<UIDropInteractionDelegate>)dropInteractionDelegate;
-- (id<BEDragInteractionDelegate>)dragInteractionDelegate;
 - (UIDropInteraction *)dropInteraction;
+#if USE(BROWSERENGINEKIT)
+- (id<BEDragInteractionDelegate>)dragInteractionDelegate;
 - (BEDragInteraction *)dragInteraction;
+#else
+- (id<UIDragInteractionDelegate>)dragInteractionDelegate;
+- (UIDragInteraction *)dragInteraction;
+#endif
 @end
 
 #endif // PLATFORM(IOS_FAMILY)

--- a/Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm
+++ b/Tools/TestWebKitAPI/ios/DragAndDropSimulatorIOS.mm
@@ -32,7 +32,6 @@
 #import "PlatformUtilities.h"
 #import "Test.h"
 #import "UIKitSPIForTesting.h"
-#import <BrowserEngineKit/BrowserEngineKit.h>
 #import <UIKit/UIDragInteraction.h>
 #import <UIKit/UIDragItem.h>
 #import <UIKit/UIDropInteraction.h>
@@ -42,6 +41,10 @@
 #import <WebKit/_WKFormInputSession.h>
 #import <wtf/RetainPtr.h>
 #import <wtf/SoftLinking.h>
+
+#if USE(BROWSERENGINEKIT)
+#import <BrowserEngineKit/BrowserEngineKit.h>
+#endif
 
 using namespace TestWebKitAPI;
 
@@ -88,29 +91,45 @@ static char dragDropSimulatorKey;
     return (id<UIDropInteractionDelegate>)self._dragDropInteractionView;
 }
 
+template<typename InteractionType>
+InteractionType *findInteractionOfType(UIView *view)
+{
+    for (id<UIInteraction> interaction in view.interactions) {
+        if (auto foundInteraction = dynamic_objc_cast<InteractionType>(interaction))
+            return foundInteraction;
+    }
+    return nil;
+}
+
+#if USE(BROWSERENGINEKIT)
+
 - (id<BEDragInteractionDelegate>)dragInteractionDelegate
 {
     return (id<BEDragInteractionDelegate>)self._dragDropInteractionView;
 }
 
-- (UIDropInteraction *)dropInteraction
+- (BEDragInteraction *)dragInteraction
 {
-    UIView *interactionView = self._dragDropInteractionView;
-    for (id <UIInteraction> interaction in interactionView.interactions) {
-        if ([interaction isKindOfClass:[UIDropInteraction class]])
-            return (UIDropInteraction *)interaction;
-    }
-    return nil;
+    return findInteractionOfType<BEDragInteraction>(self._dragDropInteractionView);
+}
+
+#else
+
+- (id<UIDragInteractionDelegate>)dragInteractionDelegate
+{
+    return (id<UIDragInteractionDelegate>)self._dragDropInteractionView;
 }
 
 - (UIDragInteraction *)dragInteraction
 {
-    UIView *interactionView = self._dragDropInteractionView;
-    for (id <UIInteraction> interaction in interactionView.interactions) {
-        if ([interaction isKindOfClass:[UIDragInteraction class]])
-            return (UIDragInteraction *)interaction;
-    }
-    return nil;
+    return findInteractionOfType<UIDragInteraction>(self._dragDropInteractionView);
+}
+
+#endif
+
+- (UIDropInteraction *)dropInteraction
+{
+    return findInteractionOfType<UIDropInteraction>(self._dragDropInteractionView);
 }
 
 @end
@@ -484,6 +503,7 @@ IGNORE_WARNINGS_END
     } else {
         _dragSession = adoptNS([[MockDragSession alloc] initWithWindow:[_webView window] allowMove:self.shouldAllowMoveOperation]);
         [_dragSession setMockLocationInWindow:_startLocation];
+#if USE(BROWSERENGINEKIT)
         [[_webView dragInteractionDelegate] dragInteraction:[_webView dragInteraction] prepareDragSession:_dragSession.get() completion:[strongSelf = retainPtr(self)] {
             if (strongSelf->_phase == DragAndDropPhaseCancelled)
                 return NO;
@@ -492,6 +512,15 @@ IGNORE_WARNINGS_END
             [strongSelf _advanceProgress];
             return YES;
         }];
+#else
+        [(id<UIDragInteractionDelegate_ForWebKitOnly>)[_webView dragInteractionDelegate] _dragInteraction:[_webView dragInteraction] prepareForSession:_dragSession.get() completion:[strongSelf = retainPtr(self)] {
+            if (strongSelf->_phase == DragAndDropPhaseCancelled)
+                return;
+
+            strongSelf->_phase = DragAndDropPhaseBeginning;
+            [strongSelf _advanceProgress];
+        }];
+#endif
     }
 
     Util::run(&_isDoneWithCurrentRun);
@@ -610,11 +639,18 @@ IGNORE_WARNINGS_END
     [_queuedAdditionalItemRequestLocations removeObjectAtIndex:0];
 
     auto requestLocation = [[_webView window] convertPoint:[requestLocationValue CGPointValue] toView:_webView.get()];
+#if USE(BROWSERENGINEKIT)
     [[_webView dragInteractionDelegate] dragInteraction:[_webView dragInteraction] itemsForAddingToSession:_dragSession.get() forTouchAtPoint:requestLocation completion:[dragSession = _dragSession, dropSession = _dropSession](NSArray *items) {
         [dragSession addItems:items];
         [dropSession addItems:items];
         return YES;
     }];
+#else
+    [(id<UIDragInteractionDelegate_ForWebKitOnly>)[_webView dragInteractionDelegate] _dragInteraction:[_webView dragInteraction] itemsForAddingToSession:_dragSession.get() withTouchAtPoint:requestLocation completion:[dragSession = _dragSession, dropSession = _dropSession] (NSArray *items) {
+        [dragSession addItems:items];
+        [dropSession addItems:items];
+    }];
+#endif
     return YES;
 }
 

